src/engine/Server.cpp

```diff
@@ -697,41 +697,40 @@ boost::asio::awaitable<void> Server::processQuery(
               << " ms" << std::endl;
     LOG(TRACE) << qet.getCacheKey() << std::endl;
 
-    // Apply stricter limit for export if present
-    if (maxSend.has_value()) {
-      auto& pq = plannedQuery.value().parsedQuery_;
-      pq._limitOffset._limit =
-          std::min(maxSend.value(), pq._limitOffset.limitOrDefault());
-    }
-    // Make sure we don't underflow here
-    AD_CORRECTNESS_CHECK(
-        plannedQuery.value().parsedQuery_._limitOffset._offset >=
-        qet.getRootOperation()->getLimit()._offset);
+    // Remember the value of the `send` parameter. It will be applied in
+    // `ExportQueryExecutionTrees` when we materialize a (part of the) result.
+    auto& limitOffset = plannedQuery.value().parsedQuery_._limitOffset;
+    limitOffset.maxSend_ = maxSend;
+    // Apply stricter limit for export if present.
+    // if (maxSend.has_value()) {
+    //   auto& pq = plannedQuery.value().parsedQuery_;
+    //   pq._limitOffset._limit =
+    //       std::min(maxSend.value(), pq._limitOffset.limitOrDefault());
+    // }
+
     // Don't apply offset twice, if the offset was not applied to the operation
     // then the exporter can safely apply it during export.
-    plannedQuery.value().parsedQuery_._limitOffset._offset -=
-        qet.getRootOperation()->getLimit()._offset;
+    //
+    // TODO<hannahbast> I don't understand why the offset would be "applied
+    // twice" without this code. Is this a code smell?
```

Comment:
All operations with a non-default limit will be cropped according to that limit if they return false for `supportsLimit()` to ensure the result size is always right.
If the limit/offset is supposed to be applied to the outermost operation this limit/offset will not get applied to that operation and instead left to the exporter to "crop" the result instead ***if the operation does return false for `supportsLimit()`***. So when `supportsLimit()` returns true, the offset would be applied twice, once by the built-in operation mechanism and once by the exporter. For the limit this doesn't matter because well a result that has been cropped to 10 elements will still contain 10 elements if cropped again to the same size. For the offset this does not work, so it has to be set to zero so the exporter doesn't skip any elements.

Reply:
Thanks for the explanation! So does this code always set `limitOffset._offset` to zero? If not, is there a case where it is decreased but not to zero?

Reply:
Yes, it will always set it to zero or not modify it at all, depending if the offset was set during query planning or not